snuba/datasets/plans/translator/visitor.py

```diff
@@ -0,0 +1,158 @@
+        default_factory=list
+    )
+    arguments: Sequence[ExpressionMapper[Argument, TExpOut]] = field(
+        default_factory=list
+    )
+    lambdas: Sequence[ExpressionMapper[Lambda, TExpOut]] = field(default_factory=list)
+
+    def concat(self, spec: TranslationRules[TExpOut]) -> TranslationRules[TExpOut]:
+        return TranslationRules(
+            literals=[*self.literals, *spec.literals],
+            columns=[*self.columns, *spec.columns],
+            subscriptables=[*self.subscriptables, *spec.subscriptables],
+            functions=[*self.functions, *spec.functions],
+            curried_functions=[*self.curried_functions, *spec.curried_functions],
+            arguments=[*self.arguments, *spec.arguments],
+            lambdas=[*self.lambdas, *spec.lambdas],
+        )
+
+
+TExpIn = TypeVar("TExpIn", bound=Expression)
+
+
+class ExpressionMapper(ABC, Generic[TExpIn, TExpOut]):
+    """
+    One translation rule used by the MappingExpressionTranslator to translate an
+    expression.
+
+    see MappingExpressionTranslator for more context
+    """
+
+    @abstractmethod
+    def attemptMap(
+        self, expression: TExpIn, children_translator: ExpressionVisitor[TExpOut]
+    ) -> Optional[TExpOut]:
+        """
+        Translates an expression if this rule matches such expression. If not, it
+        returns None.
+
+        It receives an instance of a translating visitor to take care of children
+        that this rule is not capable, or not supposed to, translate. These node
+        should be delegated to the children_translator.
+        """
+        raise NotImplementedError
+
+
+class MappingExpressionTranslator(ExpressionVisitor[TExpOut]):
+    """
+    Translates a Snuba query expression into an physical query expression (like a
+    Clickhouse query expression).
+
+    The translation of every node in the expression is performed by a series of rules
+    that extend ExpressionMapper.
+    Rules are applied in sequence. Given an expression, the first valid rule for such
+    expression is applied and the result is returned. If no rule can translate such
+    expression an exception is raised.
+    A rule can delegate the translation of its children back to this visitor.
+
+    Each rule only has context around the expression provided and its children. It does
+    not have general context around the query or around the expression's ancestors in
+    the AST.
+    This approach implies that, while rules are specific to the relationship between
+    dataset (later entities) and storage, this class keeps the responsibility of
+    orchestrating the translation process.
+
+    It is possible to compose different, independently defined, sets of rules that are
+    applied in a single pass over the AST.
+    This allows us to support joins and multi-step translations (for multi table
+    storages) as an example:
+    Joins can be supported by simply concatenating rule sets associated with each storage.
```

Comment:
I think it may be useful to give some context on how I envision joins to work since it may not be obvious:
**phase 1:** pre-entity. Here not all columns in the AST will be fully qualified. 
- non join query: Here the columns now are not fully qualified. It works as it is, columns do not reference a table alias and the rules do not care about table aliases
- join query: Here the columns are fully qualified so that they are not ambiguous. Since the translation rules do not consider table aliases we need to do soemthing about it. The Qoery Plan Builder for joins will wrap all the translation rules before applying them to account for the table alias. Something like:
```
# JoinWrapper takes in the column to translate, removes the alias, execute the inner rule
# and transform the output by adding the alias back.
storage1_rules = [JoinWrapper(rule, table_alias="event") for rule in storage1_rules]
storage2_rules = [JoinWrapper(rule, table_alias="groups") for rule in storage2_rules]
rules = [*storage1_rules, *storage2_rules]
translator = Translator(rules)
```

**phase 2:** entities.
Now we don't care about all that complexity anymore since all columns in the AST (and, as a consequences all translation rules) are fully qualified with the entity name, which replaces the table aliases in joins.